bin/cml-runner.js

```diff
@@ -5,6 +5,7 @@ const { homedir } = require('os');
 
 const fs = require('fs').promises;
 const yargs = require('yargs');
+const { SpotNotifier } = require('ec2-spot-notification');
```

Comment:
Can we separate this functionality and move it to the AMI on the terraform-provider-iterative repository? While this is a must have, keeping it as part of the CML code would introduce another unrelated responsibility to this codebase.

Reply:
We can just `curl` the `169.254.169.254` endpoint in a loop and run `systemctl stop cml` once we get a termination notice, all from the instance script.

Reply:
Lets open another ticket and use for now the already done work that also works.

Reply:
Also, we might just want this behaviour with the runner, so I guess it should go in the TPI runner resource at most

Reply:
> Can we separate this functionality and move it to the AMI and fallback on the terraform-provider-iterative repository? While this is a must have, keeping it as part of the CML code would introduce another unrelated responsibility to this codebase.

Actually I disagree, Its a part of the logic that we are setting to handle long job runners. The same that we do with the 72 hours in example. And ensures the integrty and responsability of the runner

Reply:
> _Also, we might just want this behaviour with the runner, so I guess it should go in the TPI runner resource at most_

Agreed!

Reply:
Putting it outside makes more difficult to follow up the logic that ensures this feature of the runner. It MUST be definitely inside the runner

Reply:
> _Actually I disagree, Its a part of the logic that we are setting to handle long job runners. The same that we do with the 72 hours in example. And ensures the integrty and responsability of the runner_

I also disagree in turn. 😈  This is a cloud-specific termination detection mechanism and, no matter what its purpose is, should ideally be handled at the `resource_cml_runner` level at most. This project should not care about clouds but only about SCM platforms, and use `SIGTERM` as a common signal for every termination scenario.

Reply:
> should ideally be handled at the resource_cml_runner level at most. This project should not care about clouds but only about SCM platforms, and use SIGTERM as a common signal for every termination scenario.

Why? why SCMs and not clouds? Who puts those constrains? Putting it outside its not a good idea. Indeed Im missing someting... whatch next commit

Reply:
> _Putting it outside makes more difficult to follow up the logic that ensures this feature of the runner. It MUST be definitely inside the runner_

More than a "feature" of the runner, I would consider it a cloud-vendor-specific patch to notify processes on spot termination with the same mechanism used for instance termination.

Reply:
I think that the latest commit closes the discussion

Reply:
You're making the AWS spot instance termination detection logic optional, but...

1. How are we detecting the termination of Azure spot and Google Compute preemptive instances?
2. Do you still think that it's a good idea to introduce code specific to cloud vendors in the CML repository?

Reply:
> Do you still think that it's a good idea to introduce code specific to cloud vendors in the CML repository?

yes, I would like to have all the base code together. If it were for me the TPI would be done in js... actually... I have a proposal 😉 

> How are we detecting the termination of Azure spot and Google Compute preemptive instances?

It does not exists. SIGTERM should suffice. Actually in AWS is intermittent. Seems that SIGTERM does not always work. ideally we should not need it.

Reply:
Anyway if the code works correctly and the direction is not clear is better to release and USE already developed and tested code

Reply:
> _yes, I would like to have all the base code together. If it were for me the TPI would be done in js... actually... I have a proposal_

I would too, if Hashicorp had chosen to implement Terrraform in JavaScript. Given that it's not the case, I'd prefer to separate the responsibilities where they make most sense, keeping all the cloud-specific logic in the provider and all the SCM-specific logic in CML, as we do now.

> _It does not exists. SIGTERM should suffice. Actually in AWS is intermittent. Seems that SIGTERM does not always work. ideally we should not need it._

Then, shouldn't this be a patch on the provider so we convert AWS termination notifications to a normal `SIGTERM` signal? If you can't tell the difference between spot termination and instance terminaltion in other vendors, why do it differently on this one?

Reply:
> I would too, if Hashicorp had chosen to implement Terrraform in JavaScript. Given that it's not the case,

Im thinking to do all in JS executed with V8. At the very end we are using the vendors SDK that already exists in JS

> Then, shouldn't this be a patch on the provider so we convert AWS termination notifications to a normal SIGTERM signal? If you can't tell the difference between spot termination and instance terminaltion in other vendors, why do it differently on this one?

Reasons: 

 - More complex in the TPI
 - More obscure having another service or whatever...
 - This runner could be launched manually inside an AWS instance and still works on its own

Reply:
> _Im thinking to do all in JS executed with V8. At the very end we are using the vendors SDK that already exists in JS_

Not sure about the specific advantages of V8 over any other engine for this use case, but anyhow. This is a completely separate discussion topic.

> * More complex in the TPI

Not _that_ complex.

> * More obscure having another service or whatever...

Not as obscure as moving cloud-specific code to the CML repository instead of abstracting over signals, in my opinion.

> * This runner could be launched manually inside an AWS instance and still works on its own

Do we _really_ want to support this use case? What's the advantage of supporting it if we don't do the same for the other providers?

Reply:
All advantages despite that you disagree.

Reply:
> Not as obscure as moving cloud-specific code to the CML repository

CML repository is a runner also that can check whatever is needed to ensure the feature it offers. Transparent spot instances

Reply:
> Not that complex.

We have to check also the noRetry... pass it.. create another service or whatever, business logic broken in multiple parts ... Ugly! 
I can't believe that you can not see the benefit

Reply:
> Not sure about the specific advantages of V8 over any other engine for this use case, but anyhow. This is a completely separate discussion topic.

Im telling you a solution to avoid Golang source code that Im planning to also unify the base source code that we were discussing